src/metrics/rules/active_ruleset.go

```diff
@@ -45,6 +48,69 @@ type activeRuleSet struct {
 	tagsFilterOpts  filters.TagsFilterOptions
 	newRollupIDFn   metricid.NewIDFn
 	isRollupIDFn    metricid.MatchIDFn
+
+	useFastMatch bool
+	// The key is the value of the __name__ label, the value is the mappingRule
+	mappingRuleIndex map[string][]*mappingRule
+	// The key is the value of the __name__ label, the value is rollupRule
+	rollupRuleIndex map[string][]*rollupRule
+}
+
+func newActiveRuleSetFast(
+	version int,
+	mappingRules []*mappingRule,
+	rollupRules []*rollupRule,
+	tagsFilterOpts filters.TagsFilterOptions,
+	newRollupIDFn metricid.NewIDFn,
+	isRollupIDFn metricid.MatchIDFn,
+) *activeRuleSet {
+	rset := newActiveRuleSet(version, mappingRules, rollupRules, tagsFilterOpts, newRollupIDFn, isRollupIDFn)
+	rset.useFastMatch = true
+
+	nowNanos := time.Now().UnixNano()
+
+	getNameTagValue := func(nameFilterValue *filters.FilterValue, snapshotString func() string) string {
+		if nameFilterValue == nil {
+			panic(fmt.Sprintf("A rule has no fitler for metric name tag: %s", snapshotString()))
+		}
+		if nameFilterValue.Negate {
+			panic(fmt.Sprintf("A rule has negated filter for metric name tag: %s", snapshotString()))
+		}
+		if strings.Contains(nameFilterValue.Pattern, "*") {
+			panic(fmt.Sprintf("A rule has wildcard filter for metric name tag %s", snapshotString()))
+		}
+		return nameFilterValue.Pattern
+	}
+
+	rset.mappingRuleIndex = make(map[string][]*mappingRule)
```

Comment:
just curious, based on current rule size, how much memory this index layer would need to keep right now?

Reply:
The memory footprint is tiny. There are 504 mapping rules and 554 rollup rules in `oregon-prod`. There is only one copy of the index in the process.
```
created mappingRuleIndex of 458 entries for 504 mapping rules and rollupRuleIndex of 364 entries for 554 rollup rules

```